### PR TITLE
Add linked token model exposed via graphql

### PR DESCRIPTION
**Pull request by tuntisz**
The goal here is to be able to provide users with a custom link to edit an entity. This puts us most of the way there.

When an entity (request or response) is created, we also create a LinkedToken. This token holds a reference to the entity allowing anyone that knows the token to edit it.

Query:
```
{
  linkedToken(token: "3be17075-d86e-4bec-accc-ca241abe486e") {
    ...on Request {
      provider {
        id
      }
    }
  }
}
```

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  mount GraphiQL::Rails::Engine, at: "/graphiql", graphql_path: "/graphql" if Rails.env.development?
```
Comment (cantino) [inline config/routes.rb:2]: Nice, this will be helpful. I've been using [Insomnia](https://insomnia.rest/).

File: app/services/link_creator.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class LinkCreator
+  class << self
+    def create(entity)
+      LinkedToken.create!(entity: entity, token: SecureRandom.uuid)
```
Comment (cantino) [inline app/services/link_creator.rb]: Personally, I might just add a 

```ruby
before_validation :set_token, on: :create

def set_token
  self.token ||= SecureRandom.uuid
end
```

to the model, but I'm also fine with a service.

File: app/graphql/types/linked_token_entity_union_type.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Types::LinkedTokenEntityUnionType < Types::BaseUnion
```
Comment (cantino) [inline app/graphql/types/linked_token_entity_union_type.rb:1]: Nice, I haven't used a GraphQL Union before.

File: db/schema.rb [hunk 4]
```diff
@@ -10,11 +10,20 @@
...
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2020_03_16_204230) do
+ActiveRecord::Schema.define(version: 2020_03_19_055350) do
 
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
 
+  create_table "linked_tokens", force: :cascade do |t|
+    t.string "token"
```
Comment (cantino) [inline db/schema.rb]: I think this should be `null: false`

File: db/schema.rb [hunk 5]
```diff
@@ -10,11 +10,20 @@
...
+    t.string "entity_type", null: false
+    t.bigint "entity_id", null: false
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.index ["entity_type", "entity_id"], name: "index_linked_tokens_on_entity_type_and_entity_id"
```
Comment (cantino) [inline db/schema.rb:24]: We should probably have an index on `token`.

**Review by cantino — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 1 above)
Comment (tuntisz) [reply config/routes.rb:2]: Nice, I hadn't heard of Insomnia. I'm gonna check that out

File: app/services/link_creator.rb (hunk 2 above)
Comment (tuntisz) [reply app/services/link_creator.rb]: Sweet, I didn't know this was possible. I tend to lean towards services just to limit the surface area of code using AR.

Comment (tuntisz) [reply app/services/link_creator.rb]: I updated the model to use the on create hook! Thanks for the tip.

File: db/schema.rb (hunk 5 above)
Comment (tuntisz) [reply db/schema.rb:24]: Good call. I made that change. Let me know if the added indexes make sense.

**Comment by cantino** [comment]
Nice!

**Pull request merged**